cpp/ql/test/query-tests/Security/CWE/CWE-119/semmle/tests/OverflowBuffer.expected

```diff
@@ -52,9 +52,6 @@
 | tests.cpp:577:7:577:13 | access to array | This array indexing operation accesses a negative index -1 on the $@. | tests.cpp:565:7:565:12 | buffer | array |
 | tests_restrict.c:12:2:12:7 | call to memcpy | This 'memcpy' operation accesses 2 bytes but the $@ is only 1 byte. | tests_restrict.c:7:6:7:13 | smallbuf | source buffer |
 | unions.cpp:26:2:26:7 | call to memset | This 'memset' operation accesses 200 bytes but the $@ is only 100 bytes. | unions.cpp:21:10:21:11 | mu | destination buffer |
-| unions.cpp:27:2:27:7 | call to memset | This 'memset' operation accesses 100 bytes but the $@ is only 10 bytes. | unions.cpp:15:7:15:11 | small | destination buffer |
-| unions.cpp:29:2:29:7 | call to memset | This 'memset' operation accesses 100 bytes but the $@ is only 10 bytes. | unions.cpp:15:7:15:11 | small | destination buffer |
-| unions.cpp:30:2:30:7 | call to memset | This 'memset' operation accesses 200 bytes but the $@ is only 10 bytes. | unions.cpp:15:7:15:11 | small | destination buffer |
 | unions.cpp:30:2:30:7 | call to memset | This 'memset' operation accesses 200 bytes but the $@ is only 100 bytes. | unions.cpp:15:7:15:11 | small | destination buffer |
```

Comment:
We've lost results on lines 27 and 29 here.  Both are marked `(dubious)` in the test doc (i.e. we would be happy to flag them or not flag them), and involve writing off the end of one union case but not of the union as a whole.  It's not necessarily a problem.  But I'd like to understand why the results here have changed?

Reply:
Well spotted. This has been fixed in 407a20ba12e3f82f7c8e2bc7cd8976212606d1c0. There are probably better ways of fixing this, but in the spirit of getting this merged soon I think this is okay

Reply:
I'm not much the wiser about what was going on here.

Why does `toExprString` have to work with locations?

Reply:
Sorry. Let me explain my thought process on this. The code that was missing a result was:
```cpp
union myUnion {
  char small[10];
  char large[100];
};
// ...
myUnion mu;
memset(&(mu.small), 0, sizeof(mu));
```
To raise an alert here [getBufferSize](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/commons/Buffer.qll#L80) has to conclude that `&(mu.small)` has size `10`. This should happen by:
- `isSource` concluding that `mu.small` has size 10 because it's a union
- That fact being transferred from `mu.small` to `&(mu.small)` (which is the argument of the `memset`) via the [dataflow step](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/commons/Buffer.qll#L98) in `getBufferSize`.

In order for the dataflow step to kick in, there needs to be:
1. A dataflow node `n1` such that `n1.asExpr()` is `mu.small`, and
2. another dataflow node `n2` such that `n2.asExpr()` is `&(mu.small)`
3. A dataflow step from `n1` to `n2`

The problem was that 1. wasn't satisfied. This is because `&x` in the IR is translated into code such as:
```
r1 = VariableAddress[x]
r2 = CopyValue r1
```
i.e., because `VariableAddress[x]` already represents the address of `x`, so the IR just generates a `CopyValue` instruction so that _some_ instruction maps to `&x`.

As I said in the top of this PR, this PR solves the problem of duplicated dataflow results by requiring that only _fully converted_ dataflow nodes have a result for `asExpr()`, but because the `&` operator is just another conversion (because dataflow considers `CopyValue` instructions as just another conversion that should be skipped), `r2` is actually the instruction whose `asExpr()` must return the `VariableAccess` for x. But `r2` is _also_ the instruction that contains the result of `&x` (because `r2` genuinely is the instruction that produces the result of the expression `&x`).

The above paragraph is pretty much exactly what I wrote in the large comment block above the line I modified [here](https://github.com/github/codeql/commit/407a20ba12e3f82f7c8e2bc7cd8976212606d1c0#diff-e9e4c8dbaa54c9f16e1533c09bb66ff3ec456e747f98de139672cfa0412a463cR1132) (in the commit I linked), except that the comment talks about `*x = ...` instead of `&x`.

Reply:
Alright, so with the above explanation it's hopefully clear that some dataflow nodes must map to several expressions when a `CopyValue` instruction is involved. Since we normally use `asExpr()` in code like:
```ql
exists(Expr e | e = n.asExpr() and isInteresting(e))
```
having multiple results for `asExpr()` isn't such a big deal - it's better to have two results than no result (since "no result" would mean you couldn't select the expression from a dataflow node). However, having a `toString` with multiple results is quite detrimental since it leaves you with even more duplication in your query results. But `toString()` on dataflow nodes is implemented pretty much as:
```
string toString() { result = this.asExpr().toString() }
```
(plus some more cases to handle the fact that not all dataflow nodes map to expressions)

so multiple `asExpr()` results implies multiple `toString` results, and we really don't want multiple `toString` results. So now we can finally answer:

> Why does `toExprString` have to work with locations?

Because we need a way of picking out _one_ string representation among the expressions mapped to by a dataflow node. And I just decided to pick the first expression ordered by (line number, column number).

Does that make sense?

Reply:
Makes sense.
> And I just decided to pick the first expression ordered by (line number, column number).

Assuming you don't get too many cases where you have multiple expressions at the exact same location, but with different `toString`s.  I guess with conversions filtered out, that ought not happen very often.

Reply:
Yeah, that's what I'm counting on as well 🤞